### PR TITLE
MM-14412: Enables force removals of team/channel members base on group constraints.

### PR DESCRIPTION
**Pull request by mkraft**
#### Summary

Enables force removals of team/channel members base on group constraints.
* Renames some existing methods and variables.
* Change return types to ChannelMembers and TeamMembers for some existing methods.
* Adds option to change LDAP_DATA to either 'qa' or 'test' with env variable.
* Adds methods to retrieve ChannelMembers and TeamMembers that, based on group constraints, should be deleted.
* Adds helper functions to create GroupTeams and GroupChannels.

#### Ticket Link

https://mattermost.atlassian.net/browse/MM-14412

### CODE REVIEW

File: app/syncables.go [hunk 1]
```diff
@@ -61,3 +61,48 @@ func (a *App) PopulateSyncablesSince(groupMembersCreatedAfter int64) error {
...
+		return appErr
+	}
+
+	for _, userChannel := range channelMembers {
+		channel, err := a.GetChannel(userChannel.ChannelId)
+		if err != nil {
+			return err
+		}
+
+		err = a.RemoveUserFromChannel(userChannel.UserId, "", channel)
```
Comment (mkraft) [inline app/syncables.go]: Need to verify that `removerUserId` is meant to be an empty string.

File: app/syncables.go [hunk 2]
```diff
@@ -61,3 +61,48 @@ func (a *App) PopulateSyncablesSince(groupMembersCreatedAfter int64) error {
...
+		)
+	}
+
+	teamMembers, appErr := a.TeamMembersToRemove()
+	if appErr != nil {
+		return appErr
+	}
+
+	for _, userTeam := range teamMembers {
+		err := a.RemoveUserFromTeam(userTeam.TeamId, userTeam.UserId, "")
```
Comment (mkraft) [inline app/syncables.go]: Need to verify that `requestorId` is meant to be an empty string.

File: tests/test-data.ldif [hunk 3]
```diff
@@ -127,4 +127,86 @@ objectclass: iNetOrgPerson
 sn: User
 cn: Board3
 mail: success+boardthree@simulator.amazonses.com
-userPassword: Password1
```
Comment (mkraft) [inline tests/test-data.ldif]: Everything below is an unchanged copy/paste from `tests/add-groups.ldif` it's just an amalgamation of the two files.

File: app/syncables_test.go [hunk 4]
```diff
@@ -70,32 +72,17 @@ func TestPopulateSyncablesSince(t *testing.T) {
 		t.Errorf("test group not created: %s", err.Error())
 	}
 
-	_, err = th.App.CreateGroupSyncable(&model.GroupSyncable{
-		AutoAdd:    true,
-		GroupId:    gleeGroup.Id,
-		SyncableId: practiceChannel.Id,
-		Type:       model.GroupSyncableTypeChannel,
-	})
+	_, err = th.App.CreateGroupSyncable(model.NewGroupChannel(gleeGroup.Id, practiceChannel.Id, true))
```
Comment (mkraft) [inline app/syncables_test.go]: The changes like this in this file are just switching to use the `NewGroupChannel` and `NewGroupTeam` helper functions.

File: store/sqlstore/group_supplier.go [hunk 5]
```diff
@@ -698,23 +698,23 @@ func (s *SqlSupplier) PendingAutoAddTeamMembers(ctx context.Context, since int64
 			AND (GroupMembers.CreateAt >= :Since
 			OR GroupTeams.UpdateAt >= :Since)`
 
-	var userTeamIDs []*model.UserTeamIDPair
+	var teamMembers []*model.UserTeamIDPair
 
-	_, err := s.GetMaster().Select(&userTeamIDs, sql, map[string]interface{}{"Since": since})
+	_, err := s.GetReplica().Select(&teamMembers, sql, map[string]interface{}{"Since": since})
```
Comment (mkraft) [inline store/sqlstore/group_supplier.go]: This query should have been using replica all along.

File: store/sqlstore/group_supplier.go [hunk 6]
```diff
@@ -740,14 +740,14 @@ func (s *SqlSupplier) PendingAutoAddChannelMembers(ctx context.Context, since in
 			AND (GroupMembers.CreateAt >= :Since
 			OR GroupChannels.UpdateAt >= :Since)`
 
-	var userChannelIDs []*model.UserChannelIDPair
+	var channelMembers []*model.UserChannelIDPair
 
-	_, err := s.GetMaster().Select(&userChannelIDs, sql, map[string]interface{}{"Since": since})
+	_, err := s.GetReplica().Select(&channelMembers, sql, map[string]interface{}{"Since": since})
```
Comment (mkraft) [inline store/sqlstore/group_supplier.go]: Also this query should have been using replica all along.

File: store/sqlstore/group_supplier.go [hunk 7]
```diff
@@ -765,3 +765,92 @@ func groupSyncableToGroupChannel(groupSyncable *model.GroupSyncable) *groupChann
...
+					JOIN GroupMembers ON GroupMembers.GroupId = UserGroups.Id
+				WHERE
+					Teams.GroupConstrained = TRUE
+					AND GroupTeams.DeleteAt = 0
+					AND UserGroups.DeleteAt = 0
+					AND Teams.DeleteAt = 0
+					AND GroupMembers.DeleteAt = 0
+				GROUP BY
+					Teams.Id,
+					GroupMembers.UserId)`
```
Comment (mkraft) [inline store/sqlstore/group_supplier.go]: This query (and the channel one) are the meat of this PR. 

It's getting all of the **allowed** team members (based on `GroupTeams` associations) for all of the group-constrained teams, for example:

|`TeamId`|`UserId`|
|-|-|
|`t1`|`u1`|
|`t1`|`u2`|
|`t2`|`u1`|

and comparing against the **actual** team members, for example:

|`TeamId`|`UserId`|
|-|-|
|`t1`|`u1`|
|`t1`|`u3`|

and then returning the users that are **actual** members but `NOT IN` **allowed** members:

|`TeamId`|`UserId`|
|-|-|
|`t1`|`u3`|

and that is the list of users that are "ToRemove".

File: store/sqlstore/group_supplier.go [hunk 8]
```diff
@@ -765,3 +765,92 @@ func groupSyncableToGroupChannel(groupSyncable *model.GroupSyncable) *groupChann
...
+					JOIN GroupMembers ON GroupMembers.GroupId = UserGroups.Id
+				WHERE
+					Channels.GroupConstrained = TRUE
+					AND GroupChannels.DeleteAt = 0
+					AND UserGroups.DeleteAt = 0
+					AND Channels.DeleteAt = 0
+					AND GroupMembers.DeleteAt = 0
+				GROUP BY
+					Channels.Id,
+					GroupMembers.UserId)`
```
Comment (mkraft) [inline store/sqlstore/group_supplier.go]: See https://github.com/mattermost/mattermost-server/pull/10490/files#r270404176 for an explanation of this query. Everything applies to this one too.

File: app/syncables.go [hunk 9]
```diff
@@ -27,12 +27,12 @@ func (a *App) PopulateSyncablesSince(groupMembersCreatedAfter int64) error {
...
 	}
 
-	userChannelIDs, appErr := a.PendingAutoAddChannelMembers(groupMembersCreatedAfter)
+	channelMembers, appErr := a.ChannelMembersToAdd(since)
 	if appErr != nil {
 		return appErr
 	}
 
-	for _, userChannel := range userChannelIDs {
+	for _, userChannel := range channelMembers {
```
[the remark this answers is not in the document]
Comment (mkraft) [reply app/syncables.go]: You're right and it's a good point. I have added that check. It does, however, remind me that this is only partially handling the larger issue of partial successes of syncs. My plan is to go as far as possible with the sync and then bubble-up the array of failure details to the UI. For example, "user A didn't get added to Team X", and "user B didn't get added to channel Y".

[the remark this answers is not in the document]
Comment (grundleborg) [reply app/syncables.go]: @mkraft I think your proposal for addressing the larger issue with sync partial-failures makes a lot of sense.

**Pull request merged**